wordpress-mobile/WordPress-iOS
# Refactors Publicize connection URL detection

Following on from #17803, this PR refactors the code used to detect success or failure based on the authorization URL. In particular:

* The code has now moved from the sharing web view controller to its own `PublicizeConnectionURLMatcher`
* I've added unit tests for the URL matching and associated authorize actions
* Most of the matches are now done based on `URLQueryItems` rather than just looking for substrings of the URL.

### To test

* Build and run tests and check they all pass
* Build and run and test out adding as many different accounts in My Site > Sharing as you can. I was able to test Tumblr, Twitter, and Facebook, but I don't have a LinkedIn account to test with.

## Regression Notes

1. Potential unintended areas of impact

The ability to connect any of the social accounts that are part of Publicize.

2. What I did to test those areas of impact (or what existing automated tests I relied on)

I added unit tests for the existing URL matching before I made any changes to that code, to verify that my changes didn't have any negative affect there.

3. What automated tests I added (or what prevented me from doing so)

See above. I also added extra tests around the authorize actions too. I added these at the end once I'd moved the authorize action code out of the view controller, as the VC had a bunch of dependencies making it a bit more awkward to test.

PR submission checklist:
